src/java/org/apache/cassandra/index/sai/disk/PrimaryKeyMapIterator.java

```diff
@@ -0,0 +1,103 @@
+
+import java.io.IOException;
+
+import javax.annotation.Nonnull;
+
+import org.apache.cassandra.db.DecoratedKey;
+import org.apache.cassandra.db.PartitionPosition;
+import org.apache.cassandra.dht.AbstractBounds;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.index.sai.SSTableContext;
+import org.apache.cassandra.index.sai.utils.PrimaryKey;
+import org.apache.cassandra.index.sai.utils.RangeIterator;
+
+/**
+ * Iterates keys in the PrimaryKeyMap of an SSTable.
+ * Iterating keys in the primary key map is faster than reading them from the SSTable data component
+ * because we don't deserialize all the other columns except the primary key columns.
+ * The primary key map components are also likely much smaller than the whole SSTable data component.
+ * <p>
+ * The keys are returned in token-clustering order.
+ */
+public final class PrimaryKeyMapIterator extends RangeIterator
+{
+    private final PrimaryKeyMap keys;
+    private long currentRowId;
+
+    private PrimaryKeyMapIterator(PrimaryKeyMap keys, PrimaryKey min, PrimaryKey max, long startRowId)
+    {
+        super(min, max, keys.count());
+        this.keys = keys;
+        this.currentRowId = startRowId;
+    }
+
+    public static RangeIterator create(SSTableContext ctx, AbstractBounds<PartitionPosition> keyRange) throws IOException
+    {
+        PrimaryKeyMap keys = ctx.primaryKeyMapFactory.newPerSSTablePrimaryKeyMap();
+        long count = keys.count();
+        if (keys.count() == 0)
+        {
+            keys.close();
+            return RangeIterator.empty();
+        }
+
+        PrimaryKey.Factory pkFactory = ctx.indexDescriptor.primaryKeyFactory;
+        Token minToken = keyRange.left.getToken();
+        Token maxToken = keyRange.right.getToken();
+        PrimaryKey minKeyBound = pkFactory.createTokenOnly(minToken);
+        PrimaryKey maxKeyBound = pkFactory.createTokenOnly(maxToken);
+        PrimaryKey sstableMinKey = count > 0 ? keys.primaryKeyFromRowId(0) : null;
+        PrimaryKey sstableMaxKey = count > 0 ? keys.primaryKeyFromRowId(count - 1) : null;
+        PrimaryKey minKey = (sstableMinKey == null || minKeyBound.compareTo(sstableMinKey) > 0)
+                            ? minKeyBound
+                            : sstableMinKey;
+        PrimaryKey maxKey = (sstableMaxKey == null || !maxToken.isMinimum() && maxKeyBound.compareTo(sstableMaxKey) < 0)
+                            ? maxKeyBound
+                            : sstableMaxKey;
+        assert sstableMinKey != null;
+        assert sstableMaxKey != null;
+
+        long startRowId = minToken.isMinimum() ? 0 : keys.ceiling(minKey);
+        return new PrimaryKeyMapIterator(keys, minKey, maxKey, startRowId);
+    }
+
+    @Override
+    protected void performSkipTo(PrimaryKey nextKey)
+    {
+        this.currentRowId = keys.ceiling(nextKey);
+    }
+
+    @Override
+    protected PrimaryKey computeNext()
+    {
+        return currentRowId >= 0 && currentRowId < keys.count()
```

Comment:
The updates to this class look good, though it might be helpful for future readers to add a comment indicating when this can be negative.

Reply:
Good idea! Added.